HW2/P2/parallel_vector.pyx

```diff
@@ -105,11 +115,19 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
    # to parallelize data movement.  Be sure to avoid deadlock, as well as
    # double-locking.
    ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(num_locks, locks)
+   for r in range(repeat):
+       for idx in prange(src.shape[0], nogil = True , num_threads = 4 ):
+           sidx, didx = src[idx]/N, dest[idx]/N
+           if sidx == didx:
+               acquire(&(locks[didx]))
+           else:
+               acquire(&(locks[min(didx,sidx)]))
+               acquire(&(locks[max(didx,sidx)]))
+           if counts[src[idx]] > 0:
+               counts[dest[idx]] += 1
+               counts[src[idx]] -= 1
+           if sidx == didx:
+               release(&(locks[didx]))
+           else:
+               release(&(locks[min(didx,sidx)]))
+               release(&(locks[max(didx,sidx)]))
```

Comment:
you did a good job with regard to prevent double-locking and dead-locking from occurring!